### PR TITLE
Obfuscator base implementation

### PR DESCRIPTION
**Pull request by apucher**
## Description
This PR adds a basic `Obfucator` for object trees and config values. It can integrated further with table configs and debug endpoints as needed.

The obfuscator matches key names to a configurable set of regex expressions and replaces the values with a specified masked value while serializing an arbitrary object tree. It currently supports JSON serialization only but can be extended to other formats trivially if needed.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

## Documentation
internal component only

### CODE REVIEW

**Comment by apucher** [comment]
#7232

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/Obfuscator.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    _patterns = patterns;
+  }
+
+  /**
+   * Serialize an object tree as JSON and obfuscate matching keys.
+   *
+   * @param object input value
+   * @return obfuscated JSON tree
+   */
+  public JsonNode obfuscateJson(Object object) {
```
Comment (mayankshriv) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/Obfuscator.java:91]: Should the obfuscator work at an object level or a fields/string level? One thought is that the actual obfuscator works at string level, and there's a util that takes a JSON and returns a obfuscated JSON utilizing the obfuscator.

**Review by mayankshriv — COMMENTED** [review]
Thanks a lot for finding and fixing this.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/Obfuscator.java (hunk 1 above)
Comment (apucher) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/Obfuscator.java:91]: that obfuscator would just be a set of regexes? in that case this is the "utility class" already.

I bet there's a way to extend jackson with this directly but digging into this seems overkill for what we're trying to achieve here.

**Review by mcvsubbu — COMMENTED** [review]
Will this lead to obfuscating fields that we don't want to do so? if some values (e.g. topic name, or table name etc.) contain these words?
Or ,if some subsystem (plugin) chooses to call its password as "fooBarPass", we may end up not obfuscating at all?

**Comment by apucher** [comment]
The obfuscator is fully configurable. If patterns need to added or tweaked this can be done at any time

**Comment by mcvsubbu** [comment]
> The obfuscator is fully configurable. If patterns need to be added or tweaked this can be done at any time
> 
> EDIT: to clarify - the patterns match on key name, not the values themselves

Maybe I am missing something. How is the regex configurable? Controller config?

**Comment by apucher** [comment]
> > The obfuscator is fully configurable. If patterns need to be added or tweaked this can be done at any time
> > EDIT: to clarify - the patterns match on key name, not the values themselves
> 
> Maybe I am missing something. How is the regex configurable? Controller config?

The obfuscator constructor allows injection of regexes and masking values.

This PR provides a tool - the ultimate integration will still be owned by the maintainers of the various components. Hopefully, this will get the stalled obfuscation efforts, like #7081 moving again.

**Review by mayankshriv — APPROVED** [review]

**Pull request merged**